spring-integration-file/src/main/java/org/springframework/integration/file/FileWritingMessageHandler.java

```diff
@@ -539,6 +551,11 @@ protected Object handleRequestMessage(Message<?> requestMessage) {
 					"The destination file already exists at '" + resultFile.getAbsolutePath() + "'.");
 		}
 
+		if (!exists && (FileExistsMode.APPEND.equals(this.fileExistsMode)
+				|| FileExistsMode.APPEND_NO_FLUSH.equals(this.fileExistsMode))) {
+			this.newFileCallback.handle(resultFile);
```

Comment:
This is wrong place to call this callback.
See the code like `resultFile.getParentFile().mkdirs();` below.
You told that you would like to add a header to a new file, but you need to ensure valid file system structure before you open a file for writing.

Reply:
Thanks for the correction.

Reply:
Well, looking into this one more time, I would say that we have to call this callback much further - in the `handleInputStreamMessage()` and its `WhileLockedProcessor`.
The same should be done in the `handleByteArrayMessage()`. And `handleStringMessage()`...

My- point is that we should allow to modify file only when it is locked by a single process for appending.

We might need to refactor those internal `WhileLockedProcessor` instance to be able to perform this our new `newFileCallback`

Reply:
I actually did it there the first time and got stuck on the implementation... Since state could have a `BufferedOutputStream` or `BufferedWriter` depending on the input type. Would you mind giving me an advice on how to handle this best?

Reply:
Not sure what you mean: you have that `BiConsumer<File, Message<?>>`. That's fully enough to be called from any `WhileLockedProcessor`. 
I wouldn't bother myself with its details: only the point that we can call such a `newFileCallback` from a guarded by the lock block and we definitely have to reuse the lock state.

Reply:
Oh my bad, I thought I should pass the opened writer, will change.